packages/components/src/public/ExtendedMap.ts

```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * Iterates over the ExtendedMap, calling the provided Mapper function for each key/value. Allows you to transform
+     * the values of the map.
+     * @param mapper
+     */
+    map<T>(mapper: Mapper<K, V, T>): ExtendedMap<K, T> {
+        const newMap = new ExtendedMap<K, T>();
+
+        for (const [key, value] of this) {
+            newMap.set(key, mapper(value, key, this));
+        }
+
+        return newMap;
+    }
+
+    /**
+     * Iterates over the ExtendedMap, calling the provided Reducer function for each key/value. Allows you to completely
+     * transform the ExtendedMap object into something else (e.g. a string, a filtered version of the map).
+     * @param reducer
+     * @param initialReduction
+     */
+    reduce<T>(reducer: Reducer<K, V, T>, initialReduction?: T): T {
+        let result = initialReduction === undefined ? new ExtendedMap() : initialReduction;
+
+        for (const [key, value] of this) {
+            result = reducer(result as T, value, key, this);
+        }
+
+        return result as T;
+    }
+
+    /**
+     * Creates a new ExtendedMap based on the filter function a passed in. Iterates through all of the values of the map
+     * and calls the filter function with the key, value, and whole map. If the filter function returns true we include
+     * the key/value pair in the new map.
+     * @param filterFn
+     */
+    filter(filterFn: FilterFn<K, V>): ExtendedMap<K, V> {
+        const newMap = new ExtendedMap<K, V>();
+        for (const [key, value] of this) {
+            if (filterFn(value, key, this)) newMap.set(key, value);
+        }
+        return newMap;
+    }
+
+    /**
+     * A convenience function to map over the values of the ExtendedMap. Equivalent to Array.from(myMap.values()).map().
+     * @param mapper
+     */
+    mapValues<T>(mapper: ArrayMapper<V, T>): T[] {
+        return this.valueArray.map(mapper);
+    }
+
+    find(filterFn: FilterFn<K, V>): V {
+        for (const [key, value] of this) {
+            if (filterFn(value, key, this)) return value;
+        }
+    }
+
+    /**
+     * Creates a new ExtendedMap based on this map and the map passed in as an argument.
+     * @param otherMap
+     */
+    merge(otherMap: MapType<K, V> | ExtendedMap<K, V>): ExtendedMap<K, V> {
+        return new ExtendedMap<K, V>(this, otherMap);
+    }
+
+    /**
+     * Inserts the contents of a map at the designated index. If the given index is out of range of the existing map we
+     * return a copy of the current map. If the given index is equal to the current size of the map we append the
+     * incoming otherMap.
+     * @param index: the index where to insert the otherMap
+     * @param otherMap: the otherMap to insert
+     */
+    mergeAt(index: number, otherMap: MapType<K, V> | ExtendedMap<K, V>): ExtendedMap<K, V> {
+        // Invalid, return a copy of this map
+        if (index < 0 || index > this.size) return new ExtendedMap<K, V>(this);
```

Comment:
why not just return this?

Reply:
Because it would be weird if a special incantation of this didn't return a new map. For methods like this we need to be consistent in returning a new object.

Reply:
For example, if we just returned this in this special case, then you could easily get into situations where you call `mergeAt`, then call `set` where you would be modifying the original object. Honestly I think maybe this particular case should return an error. I largely copied the logic for this method from the QueryInfo method `insertColumns` which I removed in favor of adding the logic to the ExtendedMap.